pkg/metrics/metrics.go

```diff
@@ -23,20 +23,62 @@ import (
+
+	// CreateSnapshotOperationName is the operation that tracks how long the controller takes to create a snapshot.
+	// Specifically, the operation metric is emitted based on the following timestamps:
+	// - Start_time: controller notices the first time that there is a new VolumeSnapshot CR to dynamically provision a snapshot
+	// - End_time:   controller notices that the CR has a status with CreationTime field set to be non-nil OR an error occurs first
+	CreateSnapshotOperationName = "CreateSnapshot"
+
+	// CreateSnapshotAndReadyOperationName is the operation that tracks how long the controller takes to create a snapshot and for it to be ready.
+	// Specifically, the operation metric is emitted based on the following timestamps:
+	// - Start_time: controller notices the first time that there is a new VolumeSnapshot CR(both dynamic and pre-provisioned cases)
+	// - End_time:   controller notices that the CR has a status with Ready field set to be true OR an error occurs first
+	CreateSnapshotAndReadyOperationName = "CreateSnapshotAndReady"
+
+	// DeleteSnapshotOperationName is the operation that tracks how long a snapshot deletion takes.
+	// Specifically, the operation metric is emitted based on the following timestamps:
+	// - Start_time: controller notices the first time that there is a deletion timestamp placed on the VolumeSnapshot CR and the CR is ready to be deleted. Note that if the CR is being used by a PVC for rehydration, the controller should *NOT* set the start_time.
+	// - End_time: controller removed all finalizers on the VolumeSnapshot CR such that the CR is ready to be removed in the API server.
+	DeleteSnapshotOperationName = "DeleteSnapshot"
+
+	// DynamicSnapshotType represents a snapshot that is being dynamically provisioned
+	DynamicSnapshotType = snapshotProvisionType("dynamic")
+	// PreProvisionedSnapshotType represents a snapshot that is pre-provisioned
+	PreProvisionedSnapshotType = snapshotProvisionType("pre-provisioned")
+
+	// The following metrics are used for the individual reconciliation metrics.
+	SnapshotStatusTypeUnknown            = snapshotStatusType("unknown")
+	SnapshotStatusTypeInProgress         = snapshotStatusType("in-progress")
+	SnapshotStatusTypeInvalidRequest     = snapshotStatusType("invalid-request")
+	SnapshotStatusTypeControllerError    = snapshotStatusType("controller-error")
+	SnapshotStatusTypeStorageSystemError = snapshotStatusType("storage-system-error")
+
+	// Success and Cancel are statuses for e2e operation time (operation_total_seconds)
+	// SnapshotStatusTypeSuccess represents that a CreateSnapshot, CreateSnapshotAndReady,
+	// or DeleteSnapshot has finished successfully.
+	// Individual reconciliations (reconciliation_total_seconds) also use this status.
+	SnapshotStatusTypeSuccess = snapshotStatusType("success")
+	// SnapshotStatusTypeCancel represents that a CreateSnapshot, CreateSnapshotAndReady,
+	// or DeleteSnapshot has been deleted before finishing.
```

Comment:
are we actually able to detect a canceled delete?

Reply:
The intention here was to send a cancel metric for the CreateSnapshotAndReady operation if a delete is issued prior to completion. 

In this case, we're sending a successful delete metric AND cancel CreateSnapshotAndReady metric.